MyMovies/Sources/Services/MovieService/MovieService.swift

```diff
@@ -0,0 +1,156 @@
+    ///   - apiKey: The API key for authenticating requests. Defaults to `Keys.apiKey`.
+    ///   - baseURL: The base URL for the API. Defaults to `Keys.searchURL`.
+    ///   - session: The `URLSession` instance for network requests. Defaults to `.shared`.
+    init(apiKey: String = Keys.apiKey,
+         baseURL: String = Keys.searchURL,
+         session: URLSession = .shared) {
+        self.apiKey = apiKey
+        self.baseURL = baseURL
+        self.session = session
+    }
+
+    /// Searches for movies based on a query string.
+    ///
+    /// - Parameter query: A string representing the search query.
+    /// - Returns: An array of `Movie` objects matching the query.
+    /// - Throws: Throws a `MovieError` if the request fails due to an invalid URL,
+    ///   network error, or decoding error.
+    func search(with query: String) async throws -> [Movie] {
+        do {
+            let url = try getURL(for: query)
+            let (data, _) = try await session.data(from: url)
+            let response = try JSONDecoder().decode(MovieResponse.self, from: data)
+            return response.results
+        } catch let error as DecodingError {
+            throw MovieError.decodingError(error)
+        } catch let error as URLError {
+            throw MovieError.networkError(error)
+        } catch {
+            throw error // Re-throw any unexpected errors
+        }
+    }
+
+    /// Constructs a URL for a given search query.
+    ///
+    /// - Parameter query: A string representing the search query.
+    /// - Returns: A valid `URL` for the search request.
+    /// - Throws: Throws `MovieError.invalidURL` if the URL cannot be constructed.
+    private func getURL(for query: String) throws -> URL {
+        guard let url = URL(string: "\(baseURL)/search/movie")?.appending(queryItems: [
+            URLQueryItem(name: "api_key", value: apiKey),
+            URLQueryItem(name: "query", value: query)
+        ]) else {
+            throw MovieError.invalidURL
+        }
+        return url
+    }
+}
+
+/// A mock implementation of the `Searchable` protocol for testing purposes.
+///
+/// This actor conforms to the `Searchable` protocol and provides a simple
+/// implementation of the `search` method. It returns an empty array of movies
+/// to simulate a search operation.
+///
+/// Use `MockMovieService` to test components that depend on the `Searchable` protocol
+/// without making real network calls or requiring an actual backend service.
+///
+/// ### Typealiases
+/// - `Result`: The type of the result returned by the `search` method. It is aliased to `Movie`.
+///
+/// ### Methods
+/// - `search(with:)`: Simulates a search operation and always returns an empty array.
+///
+/// Example usage:
+/// ```swift
+/// let service = MockMovieService()
+/// let movies = try await service.search(with: "query")
+/// assert(movies.isEmpty) // Always true for the mock implementation
+/// ```
+actor MockMovieService: Searchable {
```

Comment:
I kept the mock in this file purely for visibility. If this file became to long, I would have moved it into its own file.